File: src/VBNetFx/FAC_Generic/NativeTypes/MonthlyRangePickerResult.vb
Public Class MonthRangePickerResult

    Private myRangeBase As MonthRangeBase
    Private myMonthBase As RelatedMonth
    Private myFromDate As Date
    Private myToDate As Date

    Sub New()
        MyBase.New()
    End Sub

    Sub New(ByVal RangeBase As MonthRangeBase, ByVal MonthBase As RelatedMonth)
        myRangeBase = RangeBase
        myMonthBase = MonthBase
        SetDateInternal()
    End Sub

    Private Sub SetDateInternal()
        Dim currentDate As Date = Date.Now.AddMonths(-1)
        If RelatedMonth = RelatedMonth.PreviousMonth Then
            currentDate = currentDate.AddMonths(-1)
        ElseIf RelatedMonth = RelatedMonth.SecondLastMonth Then
            currentDate = currentDate.AddMonths(-2)
        End If

        myFromDate = New Date(currentDate.Year, currentDate.Month, 1)
        If MonthRangeBase = MonthRangeBase.FirstToLastPrevious Then
            myToDate = myFromDate.AddMonths(1).AddDays(-1)
        ElseIf MonthRangeBase = MonthRangeBase.FirstToLastCurrent Then
            myFromDate = myFromDate.AddMonths(1)
            myToDate = myFromDate.AddMonths(1).AddDays(-1)
        ElseIf MonthRangeBase = MonthRangeBase.FifteenthToFourteenth Then
            myFromDate = myFromDate.AddDays(14)
            myToDate = myFromDate.AddMonths(1).AddDays(-1)
        ElseIf MonthRangeBase = MonthRangeBase.TenthToNinth Then
            myFromDate = myFromDate.AddDays(9)
            myToDate = myFromDate.AddMonths(1).AddDays(-1)
        ElseIf MonthRangeBase = MonthRangeBase.TwentiethToNineteenth Then
            myFromDate = myFromDate.AddDays(19)
            myToDate = myFromDate.AddMonths(1).AddDays(-1)
        End If
    End Sub

    Public Property MonthRangeBase() As MonthRangeBase
        Get
            Return myRangeBase
        End Get
        Set(ByVal value As MonthRangeBase)
            myRangeBase = value
            SetDateInternal()
        End Set
    End Property

    Public Property RelatedMonth() As RelatedMonth
        Get
            Return myMonthBase
        End Get
        Set(ByVal value As RelatedMonth)
            myMonthBase = value
            SetDateInternal()
        End Set
    End Property

    Public Property ToDate() As Date
        Get
            Return myToDate
        End Get
        Set(ByVal value As Date)
            myToDate = value
        End Set
    End Property

    Public Property FromDate() As Date
        Get
            Return myFromDate
        End Get
        Set(ByVal value As Date)
            myFromDate = value
        End Set
    End Property
End Class

Public Enum MonthRangeBase
    FirstToLastPrevious
    FirstToLastCurrent
    TenthToNinth
    FifteenthToFourteenth
    TwentiethToNineteenth
End Enum

Public Enum RelatedMonth
    CurrentMonth
    PreviousMonth
    SecondLastMonth
End Enum

File: src/NetFxCSharp/Facesso.Generic/NativeTypes/MonthlyRangePickerResult.cs
using System;

namespace Facesso
{
    public class MonthRangePickerResult
    {
        private MonthRangeBase myRangeBase;
        private RelatedMonth myMonthBase;
        private DateTime myFromDate;
        private DateTime myToDate;

        public MonthRangePickerResult() { }

        public MonthRangePickerResult(MonthRangeBase rangeBase, RelatedMonth monthBase)
        {
            myRangeBase = rangeBase;
            myMonthBase = monthBase;
            SetDateInternal();
        }

        private void SetDateInternal()
        {
            DateTime currentDate = DateTime.Now.AddMonths(-1);
            if (RelatedMonth == RelatedMonth.PreviousMonth)
                currentDate = currentDate.AddMonths(-1);
            else if (RelatedMonth == RelatedMonth.SecondLastMonth)
                currentDate = currentDate.AddMonths(-2);

            myFromDate = new DateTime(currentDate.Year, currentDate.Month, 1);
            if (MonthRangeBase == MonthRangeBase.FirstToLastPrevious)
                myToDate = myFromDate.AddMonths(1).AddDays(-1);
            else if (MonthRangeBase == MonthRangeBase.FirstToLastCurrent)
            {
                myFromDate = myFromDate.AddMonths(1);
                myToDate = myFromDate.AddMonths(1).AddDays(-1);
            }
            else if (MonthRangeBase == MonthRangeBase.FifteenthToFourteenth)
            {
                myFromDate = myFromDate.AddDays(14);
                myToDate = myFromDate.AddMonths(1).AddDays(-1);
            }
            else if (MonthRangeBase == MonthRangeBase.TenthToNinth)
            {
                myFromDate = myFromDate.AddDays(9);
                myToDate = myFromDate.AddMonths(1).AddDays(-1);
            }
            else if (MonthRangeBase == MonthRangeBase.TwentiethToNineteenth)
            {
                myFromDate = myFromDate.AddDays(19);
                myToDate = myFromDate.AddMonths(1).AddDays(-1);
            }
        }

        public MonthRangeBase MonthRangeBase
        {
            get { return myRangeBase; }
            set { myRangeBase = value; SetDateInternal(); }
        }

        public RelatedMonth RelatedMonth
        {
            get { return myMonthBase; }
            set { myMonthBase = value; SetDateInternal(); }
        }

        public DateTime ToDate
        {
            get { return myToDate; }
            set { myToDate = value; }
        }

        public DateTime FromDate
        {
            get { return myFromDate; }
            set { myFromDate = value; }
        }
    }

    public enum MonthRangeBase
    {
        FirstToLastPrevious,
        FirstToLastCurrent,
        TenthToNinth,
        FifteenthToFourteenth,
        TwentiethToNineteenth
    }

    public enum RelatedMonth
    {
        CurrentMonth,
        PreviousMonth,
        SecondLastMonth
    }
}
